<launch>
   
  <!-- Kinect 2
      Install Kinect2 : Follow ALL directives at https://github.com/code-iai/iai_kinect2
                        Make sure it is calibrated!
      Run:
         $ roslaunch rtabmap_ros rgbd_mapping_kinect2.launch
         $ rosrun kinect2_bridge kinect2_bridge  _publish_tf:=true

      Prefixes:
         sd: rgb_lowres / depth_lowres
         hd: rgb_rect / depth_highres
         ir: ir_rect / depth_rect
  -->
  <arg name="rgb_prefix" default="qhd" />
  <arg name="depth_prefix" default="qhd" />

  <!-- Fixed frame id, you may set "base_link" or "base_footprint" if they are published -->
  <arg name="frame_id" default="kinect2_link"/>

  <!-- Choose visualization -->
  <arg name="rviz" default="false" />
  <arg name="rtabmapviz" default="true" /> 

  <!-- ODOMETRY MAIN ARGUMENTS: 
      -"strategy"        : Strategy: 0=BOW (bag-of-words) 1=Optical Flow
      -"feature"         : Feature type: 0=SURF 1=SIFT 2=ORB 3=FAST/FREAK 4=FAST/BRIEF 5=GFTT/FREAK 6=GFTT/BRIEF 7=BRISK
      -"nn"              : Nearest neighbor strategy : 0=Linear, 1=FLANN_KDTREE, 2=FLANN_LSH, 3=BRUTEFORCE 
                           Set to 1 for float descriptor like SIFT/SURF                  
                           Set to 3 for binary descriptor like ORB/FREAK/BRIEF/BRISK  
      -"max_depth"       : Maximum features depth (m)  
      -"min_inliers"     : Minimum visual correspondences to accept a transformation (m)  
      -"inlier_distance" : RANSAC maximum inliers distance (m)  
      -"local_map"       : Local map size: number of unique features to keep track 
      -"odom_info_data"  : Fill odometry info messages with inliers/outliers data.
  -->
  <arg name="strategy" default="0" />
  <arg name="feature" default="6" />
  <arg name="nn" default="3" />
  <arg name="max_depth" default="12.0" />
  <arg name="min_inliers" default="20" />
  <arg name="inlier_distance" default="0.02" />
  <arg name="local_map" default="1000" />
  <arg name="gftt_max_corners" default="500" />
  <arg name="gftt_min_distance" default="7" />

  <!-- Visualization and SLAM nodes use same data, so just subscribe once and relay messages 
  <node name="odom_relay" type="relay" pkg="topic_tools" args="/odom /odom_relay" />
  <node name="scan_relay" type="relay" pkg="topic_tools" args="/base_scan /base_scan_relay" /> -->
  <node name="camera_info_relay" type="relay" pkg="topic_tools" args="/kinect2/qhd/camera_info /kinect2/data_throttled_camera_info_relay" />
  <node name="republish_rgb" type="republish" pkg="image_transport" args="compressed in:=/kinect2/qhd/image_color_rect raw out:=/kinect2/data_throttled_image_relay" />
  <node name="republish_depth" type="republish" pkg="image_transport" args="compressed in:=/kinect2/qhd/image_depth raw out:=/kinect2/data_throttled_image_depth_relay" />

  <!-- SLAM client side -->
  <!-- args: "delete_db_on_start" and "udebug" -->
  <group ns="rtabmap">
    <node name="rtabmap" pkg="rtabmap_ros" type="rtabmap" output="screen" args="--delete_db_on_start">
          <param name="frame_id" type="string" value="kinect2_link"/>

          <remap from="rgb/image" to="/kinect2/data_throttled_image_relay"/>
          <remap from="depth/image" to="/kinect2/qhd/image_depth_rect"/>
          <remap from="rgb/camera_info" to="/kinect2/data_throttled_camera_info_relay"/>

          <param name="queue_size" type="int" value="10"/>

          <!-- RTAB-Map's parameters -->
          <param name="RGBD/PoseScanMatching" type="string" value="true"/>
          <param name="RGBD/LocalLoopDetectionSpace" type="string" value="true"/>
          <param name="LccIcp/Type" type="string" value="2"/>
          <param name="LccBow/MinInliers" type="string" value="5"/>
          <param name="LccBow/InlierDistance" type="string" value="0.1"/>
          <param name="RGBD/AngularUpdate" type="string" value="0.01"/>
          <param name="RGBD/LinearUpdate" type="string" value="0.01"/>
          <param name="Rtabmap/TimeThr" type="string" value="700"/>
          <param name="Mem/RehearsalSimilarity" type="string" value="0.45"/>
    </node>

    <!-- Grid map assembler for rviz -->
    <node pkg="rtabmap_ros" type="grid_map_assembler" name="grid_map_assembler"/>

    <!-- Visualisation RTAB-Map -->
    <node pkg="rtabmap_ros" type="rtabmapviz" name="rtabmapviz" args="-d $(find rtabmap_ros)/launch/config/rgbd_gui.ini" output="screen">
      <param name="subscribe_depth" type="bool" value="true"/>
      <param name="subscribe_odom_info" type="bool" value="true"/>
      <param name="approx_sync" type="bool" value="true"/>
      <param name="frame_id" type="string" value="kinect2_link"/>
    
      <remap from="rgb/image" to="/kinect2/data_throttled_image_relay"/>
      <remap from="depth/image" to="/kinect2/qhd/image_depth_rect"/>
      <remap from="rgb/camera_info" to="/kinect2/data_throttled_camera_info_relay"/>

    </node>

    <!-- Odometry -->
    <node pkg="rtabmap_ros" type="rgbd_odometry" name="visual_odometry" output="screen">
      <remap from="rgb/image" to="/kinect2/data_throttled_image_relay"/>
      <remap from="depth/image" to="/kinect2/qhd/image_depth_rect"/>
      <remap from="rgb/camera_info" to="/kinect2/data_throttled_camera_info_relay"/>
      
      <param name="frame_id" type="string" value="$(arg frame_id)"/>
      <param name="approx_sync" type="bool" value="true"/>
    
      <param name="Odom/Strategy" type="string" value="$(arg strategy)"/> 
      <param name="Odom/FeatureType" type="string" value="$(arg feature)"/>  
      <param name="OdomBow/NNType" type="string" value="$(arg nn)"/>
      <param name="Odom/MaxDepth" type="string" value="$(arg max_depth)"/>  
      <param name="Odom/MinInliers" type="string" value="$(arg min_inliers)"/> 
      <param name="Odom/InlierDistance" type="string" value="$(arg inlier_distance)"/>       
      <param name="OdomBow/LocalHistorySize" type="string" value="$(arg local_map)"/>
      <param name="Odom/FillInfoData" type="string" value="$(arg rtabmapviz)"/>
      <param name="GFTT/MaxCorners" type="string" value="$(arg gftt_max_corners)"/>
      <param name="GFTT/MinDistance" type="string" value="$(arg gftt_min_distance)"/>
    </node>
  </group>

  <!-- Visualization RVIZ -->
  <node if="$(arg rviz)" pkg="rviz" type="rviz" name="rviz" args="-d $(find rtabmap_ros)/launch/config/rgbd.rviz"/>
  <!-- sync cloud with odometry and voxelize the point cloud (for fast visualization in rviz) -->
  <node if="$(arg rviz)" pkg="nodelet" type="nodelet" name="standalone_nodelet"  args="manager" output="screen"/>
  <node if="$(arg rviz)" pkg="nodelet" type="nodelet" name="data_odom_sync" args="load rtabmap_ros/data_odom_sync standalone_nodelet">
    <remap from="rgb/image_in"       to="/kinect2/data_throttled_image_relay"/>
    <remap from="depth/image_in"     to="/kinect2/qhd/image_depth_rect"/>
    <remap from="rgb/camera_info_in" to="/kinect2/data_throttled_camera_info_relay"/>

    <remap from="odom_in"             to="rtabmap/odom"/>

    <param name="approx_sync" type="bool" value="true"/>
    
    <remap from="rgb/image_out"       to="data_odom_sync/image"/>
    <remap from="depth/image_out"     to="data_odom_sync/depth"/>
    <remap from="rgb/camera_info_out" to="data_odom_sync/camera_info"/>
    <remap from="odom_out"            to="odom_sync"/>
  </node>
  <node if="$(arg rviz)" pkg="nodelet" type="nodelet" name="points_xyzrgb" args="load rtabmap_ros/point_cloud_xyzrgb standalone_nodelet">
    <remap from="rgb/image"       to="data_odom_sync/image"/>
    <remap from="depth/image"     to="data_odom_sync/depth"/>
    <remap from="rgb/camera_info" to="data_odom_sync/camera_info"/>
    <remap from="cloud"           to="voxel_cloud" />

    <param name="voxel_size" type="double" value="0.01"/>
  </node>

</launch>